test/e2e/cluster_upgrade.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Skipped", func() {
+	Describe("Cluster upgrade", func() {
+		svcName := "baz"
+		var podName string
+		framework := Framework{BaseName: "cluster-upgrade"}
+		var webserver *WebserverTest
+
+		BeforeEach(func() {
+			framework.beforeEach()
+			webserver = NewWebserverTest(framework.Client, framework.Namespace.Name, svcName)
+			pod := webserver.CreateWebserverPod()
+			podName = pod.Name
+			svc := webserver.BuildServiceSpec()
+			svc.Spec.Type = api.ServiceTypeLoadBalancer
+			webserver.CreateService(svc)
+		})
+
+		AfterEach(func() {
+			framework.afterEach()
+			webserver.Cleanup()
+		})
+
+		It("by kube-push of master should maintain responsive services", func() {
+			testClusterUpgrade(framework, svcName, podName, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-push.sh", "-m")
+			})
+		})
+
+		It("by gce-upgrade of master should maintain responsive services", func() {
+			if !providerIs("gce") {
+				By(fmt.Sprintf("Skippingt test, which is not implemented for %s", testContext.Provider))
+				return
+			}
+			testClusterUpgrade(framework, svcName, podName, func() {
+				runUpgradeScript("hack/e2e-internal/e2e-upgrade.sh", "-M", "-l")
+			})
+		})
+	})
+})
+
+func testClusterUpgrade(framework Framework, svcName, podName string, upgrade func()) {
+	result, err := waitForLoadBalancerIngress(framework.Client, svcName, framework.Namespace.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if len(result.Status.LoadBalancer.Ingress) != 1 {
```

Comment:
optional nits:

- for super easy visual numerical diffing, I'd replace "one" with "1"
- `result.Status.LoadBalancer.Ingress` is long enough I'd do something like `in := result.Status.LoadBalancer.Ingress` before the `if` check and then use that variable the four times you use it.